pydatastructs/trees/binary_trees.py

```diff
@@ -623,7 +627,197 @@ def _left_rotate(self, j, k):
         kp = self.tree[k].parent
         if kp is None:
             self.root_idx = k
+            
+class CartesianTree(SelfBalancingBinaryTree):
 
+    """
+    Represents cartesian trees.
+
+    Examples
+    ========
+
+    >>> from pydatastructs.trees import CartesianTree as CT
+    >>> c = CT()
+    >>> c.insert(1, 1, 4)
+    >>> c.insert(2, 2, 3)
+    >>> child = b.tree[b.root_idx].right
+    >>> c.tree[child].data
+    1
+    >>> c.search(1)
+    0
+    >>> c.search(-1) is None
+    True
+    >>> c.delete(1) is True
+    True
+    >>> c.search(1) is None
+    True
+    >>> c.delete(2) is True
+    True
+    >>> c.search(2) is None
+    True
+
+    References
+    ==========
+
+    .. [1] https://www.cs.princeton.edu/courses/archive/spr09/cos423/Lectures/geo-st.pdf
+
+    See Also
+    ========
+
+    pydatastructs.trees.binary_tree.SelfBalancingBinaryTree
+    """
+    
+    def _left_rotate(self, j, k):
+        y = self.tree[k].left
+        if y is not None:
+            self.tree[y].parent = j
+        self.tree[j].right = y
+        self.tree[k].parent = self.tree[j].parent
+        if self.tree[k].parent is not None:
+            if self.tree[self.tree[k].parent].left == j:
+                self.tree[self.tree[k].parent].left = k
+            else:
+                self.tree[self.tree[k].parent].right = k
+                
+        self.tree[j].parent = k
+        self.tree[k].left = j
+        kp = self.tree[k].parent
+        if kp is None:
+            self.root_idx = k
+            
+    def _right_rotate(self, j, k):
+        y = self.tree[k].right
+        if y is not None:
+            self.tree[y].parent = j
+        self.tree[j].left = y
+        self.tree[k].parent = self.tree[j].parent
+        if self.tree[k].parent is not None:
+            if self.tree[self.tree[k].parent].left == j:
+                self.tree[self.tree[k].parent].left = k
+            else:
+                self.tree[self.tree[k].parent].right = k
+        self.tree[j].parent = k
+        self.tree[k].right = j
+        kp = self.tree[k].parent
+        if kp is None:
+            self.root_idx = k
```

Comment:
Why overrided here? Are the logics for `_left_rotate` and `_right_rotate` different for AVLTree and CartesianTree?

Reply:
Can you try commenting them out and then observing the outputs?

Reply:
I was also having hard time understanding rotations implemented in SelfBalancingTree.